### ISSUE TITLE
topk over a pow2 stick-multiple width (64, 128) crashes the DDC scheduler ("Incorrect chunk size")

### ISSUE THREAD

**Issue by ani300** [torch-spyre/torch-spyre#3768]
## Summary

The all-device Gemma-4 MoE router runs `torch.topk(probs, K, dim=-1)` over the expert dimension E. At **E=128** (and E=64) the backend aborts in the DDC scheduler with:

```
DtException: Incorrect chunk size in bytes.   L3DlOpsScheduler.cpp:1714
sbf-run-scheduler-on-sdsc: failed on program 'sdsc_..._topk...'
```

The design-space scheduler derives an on-chip buffer chunk whose byte-capacity is not a whole multiple of 128 B (one stick) for the topk program.

## Not monotonic in width — the pow2 stick-multiples fail

Width sweep on the topk input (softmax → pad → topk):

| Width | Result |
|---|---|
| 32  | RUNS |
| 64  | **ABORT** |
| 96  | RUNS |
| 128 | **ABORT** |
| 130 | RUNS |
| 160 | RUNS |
| 192 | RUNS |

The failing widths 64 and 128 are the ones that are both powers of two **and** stick-multiples; non-pow2 widths (96/160/192) lower fine. The topk backend appears to pick a binary-tree tiling that only misaligns the chunk at pow2 widths.

Isolation also confirms it is topk specifically: `softmax` over `[T,128]` lowers fine; adding `topk` is the first failing op.

## Why it matters to MoE

The router's natural expert count is a power of two (Gemma-4: E=128). Without a workaround there is no clean `E = power-of-two` router path on device.

## Workaround currently in the adapter

Pad the router logits `[T,128] → [T,160]` (non-pow2 width) with `-30000.0` (≈ −∞ fp16) before topk; the pad columns never win top-K, and only the **kth value** is used as a threshold over the original `[T,128]` probs (the fp16 index is never materialized — that avoids a separate fp16→int32 CPU fallback). Costs an extra pad + concat per router call.

## Relationship to existing issues

Distinct from #3636 (topk *shape/dtype* limits: single-row `T=1`, `k>4`, rank-3, `largest=False`). Those are unsupported shapes/args; this is a **buffer-chunk alignment crash at otherwise-valid shapes**, triggered purely by the input width being a pow2 stick-multiple.

Epic: #3565

**Comment by lupalby** [comment]
could you provide a small reproducer or a bundle?

**Comment by ani300** [comment]
Here's a small standalone reproducer (no model load — just `softmax`/`topk` on a 2-D fp16 tensor). Testing on a recent branch (`adnan/ah/dense-hint-minimal-pr3890`, tip `614f9dbb`, which includes #3782 and #3874) sharpens the picture:

- **Isolated `torch.topk`** (softmax applied *outside* the compiled region) now **RUNS** at every width I tried, including the pow2 stick-multiples 64 and 128 called out in the report. #3782 ("extend work_division to split topk k across cores") and/or #3874 ("Fixed relayout on topk") seem to have addressed the isolated case.
- **Fused `softmax` + `topk`** (both inside the compiled graph — the real router shape) **still ABORTS** — and now at *every* width including the non-pow2 **160**, which this issue used as the workaround. So once the softmax fuses into the topk graph, the non-pow2 escape hatch no longer holds and the failure is no longer confined to pow2 widths.

| case | 64 | 96 | 128 | 160 | 192 |
|---|---|---|---|---|---|
| `topk` only (softmax outside compile) | RUNS | RUNS | RUNS | RUNS | RUNS |
| `softmax` + `topk` fused | **ABORTS** | **ABORTS** | **ABORTS** | **ABORTS** | **ABORTS** |

The fused case fails with the same DDC error as originally reported:

```
loc("…/sdsc_fused__softmax_topk_0_…/bundle.mlir":263:3): error: sbf-ddc: DtException: Incorrect chunk size in bytes., file …/dcg/dcg_fe/scheduler/L3DlOpsScheduler.cpp line 1714
loc(…): error: sbf-run-scheduler-on-sdsc: failed on program 'sdsc_6'
dbo: pipeline failed for '…/bundle.mlir'
```

Reproducer:

```python
import torch  # torch first; importing torch_spyre alone circular-imports

T, K, DTYPE = 64, 4, torch.float16
WIDTHS = [64, 96, 128, 160, 192]

def _topk_only(x):                      # softmax done outside compile
    return torch.topk(x, K, dim=-1)

def _softmax_then_topk(x):              # the router pattern, fused
    return torch.topk(torch.softmax(x, dim=-1), K, dim=-1)

def run(fn, x):
    try:
        v, i = torch.compile(fn, dynamic=False)(x)
        _ = v.cpu(), i.cpu()            # force materialization
        return "RUNS"
    except Exception as e:
        return f"ABORTS ({str(e).splitlines()[0]})"

for w in WIDTHS:
    pre = torch.softmax(torch.randn(T, w, dtype=DTYPE), dim=-1).to(DTYPE).to("spyre")
    print(f"w={w:4d}  topk-only     : {run(_topk_only, pre)}")
    raw = torch.randn(T, w, dtype=DTYPE).to("spyre")
    print(f"w={w:4d}  softmax+topk  : {run(_softmax_then_topk, raw)}")
```

The DDC error is emitted by the `dxp_standalone` subprocess to its own stderr (the Python side only sees the outer `CalledProcessError` naming the bundle dir); rerun `dxp_standalone -d <bundle_dir>` on a failing case to see the `line 1714` fault directly.

**Comment by lupalby** [comment]
Root-caused this; the picture on current `main` differs from the original report in two ways worth recording.

**The width pattern is stale.** The pow2/stick-multiple failures (64 and 128 aborting while 96/130/160/192 ran) no longer reproduce — isolated `topk` passes at every width on `main`, with or without any change, so #3782/#3874/#3890 covered that. The DDC `Incorrect chunk size in bytes.` (`L3DlOpsScheduler.cpp:1714`) is also gone: a frontend `OpSpecValidation` check added since the reporting branch now rejects the bad OpSpec before deeptools sees it, so the same graph fails earlier and in Python.

**What remains is width-independent and in the layout pass.** `_topk_layouts` never constrained topk's *input* layout — it attached `AllSameNode`, which forces input and output to share a stick, when topk in fact requires them to differ (the reduction dim must not be innermost). Isolated `topk` reads a graph input, whose layout is chosen freely and happens to be legal. Fused `softmax`→`topk` inherits softmax's reduction-on-stick layout and fails validation. That is why the minimal reproducer showed `topk`-only passing and `softmax`+`topk` failing at every width.

Fix in #4355: pin the required input layout and schedule the restickify via `FixedInOutNode`, mirroring `_keep_by_index_layouts`, which already did this correctly.

**On the workaround.** The `[T,128]→[T,160]` `-30000.0` pad described here isn't in `hf-adapters` anywhere I can see: not on `main`, not in history, not on any pushed branch, and not in any open or closed PR. Looks like it stayed local, so whoever has it will need to drop it themselves once #4355 lands.

Worth noting the current `hf-adapters` prefill router already compiles on `main` without #4355, because it routes topk's *indices* into `keep_by_index`, whose own input pinning propagates a legal layout back through the graph. The bug bites when topk's *values* are consumed with no `keep_by_index` present:

| graph | before | after |
|---|---|---|
| `softmax` → `topk` values | ABORTS | RUNS |
| `softmax` → `topk` indices → `keep_by_index` | RUNS | RUNS |

So #4355 unblocks the general fused pattern rather than un-breaking today's adapter.

**Comment by lupalby** [comment]
@ani300 could you please remove [backend-compiler] from the issue title? I can't edit it, and we already have the label for that.

**Comment by vswagath1989** [comment]
@ani300 - can we close this issue? and open a new one with the error we are hitting in the moment ? There is no more DDC crash. thanks

**Comment by avery-blanchard** [comment]
This was fixed in #4694, right @marnold-ibm ?

**Comment by dgrove-oss** [comment]
Confirmed fix at eed42db682ea26ec4b7773d7298a341aa048fd30

```
(torch-spyre) -bash-5.2$ python issue_3768.py 
w=  64  topk-only     : RUNS
w=  64  softmax+topk  : RUNS
w=  96  topk-only     : RUNS
w=  96  softmax+topk  : RUNS
w= 128  topk-only     : RUNS
w= 128  softmax+topk  : RUNS
w= 160  topk-only     : RUNS
w= 160  softmax+topk  : RUNS
w= 192  topk-only     : RUNS
w= 192  softmax+topk  : RUNS
```
